### PR TITLE
Use throw rather than abort() for guid parse failures

### PR DESCRIPTION
**Pull request by dfields-msft**
Make `winrt::guid("...")` more useful at runtime by throwing on failure instead of aborting the program. See #991

### CODE REVIEW

**Comment by dfields-msft** [comment]
@kennykerr does a successful run of "C++ for WinRT Xlang Internal Build" include the Clang validation you were asking for?

**Review by kennykerr — CHANGES_REQUESTED** [review]
Thanks! Please add a test that verifies this works. At a glance, this won't compile.

File: strings/base_types.h [hunk 1]
```diff
@@ -89,7 +89,7 @@ WINRT_EXPORT namespace winrt
         {
```
Comment (kennykerr) [inline strings/base_types.h:89]: This function can't be `noexcept` otherwise it will just abort as before.

Also, all of the `hex_to_xxx` functions are noexcept and terminate on failure. They would need to be updated as well.

**Comment by kennykerr** [comment]
Yes, you can either validate it locally or run the build once a test has been added.

**Comment by kennykerr** [comment]
Oh and the CI build for some reason will only run branches local to this repo.

**Comment by dfields-msft** [comment]
Does the test infrastructure have any existing support for compile-time negative tests (i.e. code that's expected to _not_ compile)? I can certainly add compile-time and runtime positive tests, as well as a runtime negative test, but I'm not going to bite off setting up test infrastructure for compile-time negative tests if that's not already available...

**Comment by kennykerr** [comment]
No negative testing unfortunately.

Just need a test that validates it works in a constant expression (e.g. `static_assert`) and a runtime test that validates it works successfully (e.g. `assert`) and unsuccessfully (e.g. throw and catch).

File: strings/base_types.h [hunk 2]
```diff
@@ -170,16 +170,27 @@ WINRT_EXPORT namespace winrt
...
+            (left.Data3 == right.Data3) &&
+            (left.Data4[0] == right.Data4[0]) &&
+            (left.Data4[1] == right.Data4[1]) &&
+            (left.Data4[2] == right.Data4[2]) &&
+            (left.Data4[3] == right.Data4[3]) &&
+            (left.Data4[4] == right.Data4[4]) &&
+            (left.Data4[5] == right.Data4[5]) &&
+            (left.Data4[6] == right.Data4[6]) &&
+            (left.Data4[7] == right.Data4[7]);
     }
```
Comment (kennykerr) [inline strings/base_types.h:187]: Hmm, I'm worried this may have an impact on code size. `memcmp` was carefully picked to reduce code size as it is used repeatedly in every QueryInterface implementation. We'd have to build something big (like WinUI) with LTCG to see whether it remains an issue.

Comment (kennykerr) [reply strings/base_types.h:187]: @DefaultRyan may have some more recent experience with this. My data is pretty old.

Comment (dfields-msft) [reply strings/base_types.h:187]: Ah, ok; I wasn't aware of those considerations. On the one hand, I'm curious and hopeful that the compiler can be smart about this esp. since it's `constexpr`. On the other hand, the motivation is a it weak; I only changed this to make the test easier to write. So it's probably best to revert this part and hack the test together instead.

Comment (sylveon) [reply strings/base_types.h:187]: I'd recommend `if (std::is_constant_evaluated())` for this, but then that's C++20 only.

Comment (sylveon) [reply strings/base_types.h:187]: Just tested if the compiler was smart about it, it isn't: compiling with `/FA` gives this

image

Comment (dfields-msft) [reply strings/base_types.h:187]: Thanks for looking into it! I'll revert this part of the change for now.

File: strings/base_types.h [hunk 3]
```diff
@@ -36,22 +36,22 @@ namespace winrt::impl
         }
         else 
         {
-            abort();
+            throw std::invalid_argument("Character is not a hexadecimal digit");
```
Comment (sylveon) [inline strings/base_types.h:39]: Can we put the character in the error message?

File: strings/base_types.h [hunk 4]
```diff
@@ -85,11 +85,11 @@ WINRT_EXPORT namespace winrt
     private:
 
         template <typename TStringView>
-        static constexpr guid parse(TStringView const value) noexcept
+        static constexpr guid parse(TStringView const value)
         {
             if (value.size() != 36 || value[8] != '-' || value[13] != '-' || value[18] != '-' || value[23] != '-')
             {
-                abort();
+                throw std::invalid_argument("value is not a valid GUID string");
```
Comment (sylveon) [inline strings/base_types.h:92]: Same here, put the actual value rather than just `value`?

File: strings/base_types.h (hunk 3 above)
Comment (dfields-msft) [reply strings/base_types.h:39]: The minor complication is that it could be a `char` or a `wchar_t`... likewise for the string in your other comment.

Comment (sylveon) [reply strings/base_types.h:39]: Ah, fair enough. Basic ASCII for char and wchar_t is the same, but if random garbage ends up being passed, this might not be desirable

**Comment by dfields-msft** [comment]
@kennykerr I'm not sure how to trigger the "C++ for WinRT Xlang internal build" pipeline run that's required by policy... Is this even possible, since my changes are coming from a different fork? I don't think I have permission to create a branch in the official microsoft fork.

**Comment by kennykerr** [comment]
I've given you Write access so you should be able to push a branch. Then you can go here and build your branch:

https://dev.azure.com/microsoft/Dart/_build?definitionId=31784

**Comment by dfields-msft** [comment]
Verified offline that Clang works with these changes too. @kennykerr any additional changes needed, or is this ready to merge?

**Review by kennykerr — APPROVED** [review]

**Comment by dfields-msft** [comment]
Thanks for approving the changes, @kennykerr! Looks like I'm not authorized to merge, so you may need to complete this PR yourself...

**Comment by kennykerr** [comment]
@DefaultRyan or @jlaanstra can you help? I'm on the beach. 😎
